### PR TITLE
Add `zenml stack export` and `zenml stack import` commands

### PR DESCRIPTION
**Pull request by fa9r**
## Describe changes
I added two CLI commands for exporting/importing ZenML stacks, which allow us to quickly transfer stacks between profiles (or even machines).

Usage:
- "zenml stack export <file_name.yaml> <stack_name>"
- "zenml stack import <file_name.yaml> <stack_name>"

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [x] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/features/integrations) table.
- [x] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

**Note:** This is a copy of [this PR](https://github.com/zenml-io/zenml/pull/554), which was automatically closed due to branch renaming. Previous review comments were resolved, see comments on previous PR for details.

### CODE REVIEW

File: src/zenml/cli/stack.py [hunk 1]
```diff
@@ -681,3 +688,125 @@ def down_stack(force: bool = False) -> None:
...
+    stack_name = stack_name or active_stack_name
+
+    if not stack_name:
+        cli_utils.warning("No stack is set as active!")
+        return
+
+    stack_configurations = repo.stack_configurations
+    if len(stack_configurations) == 0:
+        cli_utils.warning("No stacks registered!")
+        return
```
Comment (strickvl) [inline src/zenml/cli/stack.py:713]: Not sure but wonder if we should instead have a `cli_utils.error` here instead?

File: src/zenml/cli/stack.py [hunk 2]
```diff
@@ -681,3 +688,125 @@ def down_stack(force: bool = False) -> None:
...
+
+    try:
+        stack_configuration = stack_configurations[stack_name]
+    except KeyError:
+        cli_utils.error(f"Stack '{stack_name}' does not exist.")
+        return
```
Comment (strickvl) [inline src/zenml/cli/stack.py:719]: Is this necessary? The `cli_utils.error` raises an exception.

File: tests/unit/cli/test_stack.py [hunk 3]
```diff
@@ -220,3 +226,59 @@ def test_deleting_stack_with_flag_succeeds(clean_repo) -> None:
...
+    runner = CliRunner()
+    result = runner.invoke(export_stack, ["default.yaml", "default"])
+    assert result.exit_code == 0
+    assert os.path.exists("default.yaml")
+
+
+def test_stack_export_delete_import(clean_repo) -> None:
+    """Test exporting, deleting, then importing a stack succeeds."""
+    # create new stack
+    artifact_store_name = "arias_artifact_store"
```
Comment (strickvl) [inline tests/unit/cli/test_stack.py:242]: 🐈‍⬛

**Review by strickvl — CHANGES_REQUESTED** [review]
Really love this new functionality. Should help a ton with debugging.

I had two usability comments:

- The arguments being passed in when exporting somehow don't quite make sense to me. Intuitively, when I was trying this out on my machine, when I wanted to export a single stack I would want to type `zenml stack export local_stack` (where the specific stack to be exported was called `local_stack`. This actually exported *the default stack* and saved it with the filename `local_stack`. Similarly when you type `zenml stack export` on its own without arguments, I would have expected that to export everything, but it seems that it exports just the active stack. This might be something for the `#feedback` Discord channel to ask what people thing should be the behaviour, but IMO the current implementation feels a bit counterintuitive (esp as some of these behaviours as described above aren't really documented in the help strings / docstrings).
- So after I did the `zenml stack export local_stack` above, I went ahead, created a new profile, set it as my active profile and did `zenml stack import local_stack`. (The file that I was importing included a `default` stack since that was exported above.) I then saw that it asked me to choose a new name for the stack and then all the components. I recognise that the `path` is going to be different when comparing stacks, but could we enable something such that if the stack component name and attributes are identical between stack components then we don't need to create new ones under a new name? I think we'd only need to ignore the `path` attribute since that's obviously going to be different, but if everything else is the same they should be identical and thus don't need to be imported or created new.

The second point in particular is a bit of legwork to get it working but I think it'll improve the user experience, since everyone is always going to have those `default` components in every new profile they create.

File: src/zenml/cli/stack.py [hunk 4]
```diff
@@ -681,3 +688,125 @@ def down_stack(force: bool = False) -> None:
...
+    data = read_yaml(filename)
+
+    # assert zenml version is the same
+    if data["zenml_version"] != zenml.__version__:
+        cli_utils.error(
+            f"Cannot import stacks from other ZenML versions. "
+            f"The stack was created using ZenML version {data['zenml_version']}, "
+            f"you have version {zenml.__version__} installed."
+        )
+        return
```
Comment (schustmi) [inline src/zenml/cli/stack.py:764]: Same as @strickvl s comment above, the return can be removed as `cli_utils.error` is raising an exception

**Review by schustmi — APPROVED** [review]
Nice! I pointed out another superfluous return statement (which doesn't affect the functionality at all but would probably be nice if we get rid of it) so I'll approve it already

File: src/zenml/cli/stack.py (hunk 1 above)
Comment (fa9r) [reply src/zenml/cli/stack.py:713]: You're totally right. I copy-pasted that from another command, but warning()+return() doesn't make much sense IMO. Changed it in my code as well as all other places in stacks.py.

File: tests/unit/cli/test_stack.py (hunk 3 above)
Comment (fa9r) [reply tests/unit/cli/test_stack.py:242]: important to stick to the naming convention ;)

Comment (fa9r) [reply tests/unit/cli/test_stack.py:242]: 🐈‍⬛

File: src/zenml/cli/stack.py (hunk 2 above)
Comment (fa9r) [reply src/zenml/cli/stack.py:719]: True, removed the redundant return, thanks!

File: src/zenml/cli/stack.py (hunk 4 above)
Comment (fa9r) [reply src/zenml/cli/stack.py:764]: Yep, just removed it, thanks!

**Comment by fa9r** [comment]
@strickvl great suggestions! Regarding your two major points:
- syntax for both calls is now `zenml stack (import | export) <stackname> [<file.yaml>]` . Calling export without stackname no longer works (we could add this back in later if requested).
- when importing a component, if the component config of an existing component is exactly the same then it will reuse that component. However, I did not ignore the path when comparing, which forces local artifact stores and metadata stores to be recreated no matter what.

File: src/zenml/cli/stack.py [hunk 5]
```diff
@@ -681,3 +683,158 @@ def down_stack(force: bool = False) -> None:
...
+        filename = stack_name
+        data = read_yaml(filename)
+        stack_name = data["stack_name"]  # read stack_name from export
+
+    # standard 'zenml stack import stack_name [file.yaml]' calls
+    else:
+        # if filename is not given, assume default export name "<stack_name>.yaml"
+        if filename is None:
+            filename = stack_name + ".yaml"
+        data = read_yaml(filename)
```
Comment (strickvl) [inline src/zenml/cli/stack.py:810]:
```suggestion
        data = read_yaml(filename)
        cli_utils.declare(f"Using '{filename}' to import '{stack_name}' stack.")
```

Just something small like this?

**Review by strickvl — CHANGES_REQUESTED** [review]
@fa9r I've noticed that you made it so that if someone does `zenml stack import local_stack` it'll look for `local_stack.yaml` and import that. My only remaining request would be that you could add a logged message stating that it was `local_stack.yaml` that was imported instead of just silently doing that? Otherwise looks great / good to go from my side.

**Review by strickvl — APPROVED** [review]
Approved, but I think you have some conflicts to resolve before you can merge into `develop`.

**Review by strickvl — CHANGES_REQUESTED** [review]
One final thing I just remembered. Can you add the relevant documentation for these commands in the `src/zenml/cli/__init__.py` mega-docstring?

**Comment by fa9r** [comment]
> One final thing I just remembered. Can you add the relevant documentation for these commands in the `src/zenml/cli/__init__.py` mega-docstring?

Added a description of the commands.

**Pull request merged**